pkgs/applications/version-management/forgejo/default.nix

```
@@ -1,16 +1,32 @@
-{ fetchurl, gitea, lib }:
+{ lib, fetchurl, gitea
+, pamSupport ? true
+, sqliteSupport ? true
+}:
 
 gitea.overrideAttrs (old: rec {
   pname = "forgejo";
-  version = "1.18.0-rc1-1";
+  version = "1.18.0-1";
 
   src = fetchurl {
     name = "${pname}-src-${version}.tar.gz";
     # see https://codeberg.org/forgejo/forgejo/releases
-    url = "https://codeberg.org/attachments/976c426a-3e04-49ff-9762-47fab50624a3";
-    hash = "sha256-kreBMHlMVB1UeG67zMbszGrgjaROateCRswH7GrKnEw=";
+    url = "https://codeberg.org/attachments/c829784c-3b85-4996-9dc6-09e12e40a93a";
+    hash = "sha256-42bR18T5ATVyhPejeHpOTkeP6VwYu5HFfei8oMjRJy8=";
   };
 
+  preBuild =
+    let
+      tags = lib.optional pamSupport "pam"
+        ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
+      tagsString = lib.concatStringsSep " " tags;
+    in
+    ''
+      export buildFlagsArray=(
+        -tags="${tagsString}"
+        -ldflags='-X "main.Version=${version}" -X "main.Tags=${tagsString}"'
+      )
+    '';
```

`ldflags` and `tags` are canonically provided via attr:
```suggestion
ldflags = [
  "-X main.Version=${version}"
  "-X main.Tags=${lib.concatStringsSep " " tags}"
];

tags = lib.optional pamSupport [ "pam" ] ++ lib.optional sqliteSupport [ "sqlite" "sqlite_unlock_notify" ];
```

---

Error build:
```
error: cannot coerce a list to a string

           20|     "-X main.Version=${version}"
           21|     "-X main.Tags=${lib.concatStringsSep " " tags}"
             |                     ^
           22|   ];
(use '--show-trace' to show detailed location information)

```

---

You want to use lib.optional**s** with lists

---

Did you try to use the correct type when concatenating?

---

Haven't tried. Any possible examples?

---

This variant not working:
```diff
diff --git a/pkgs/applications/version-management/forgejo/default.nix b/pkgs/applications/version-management/forgejo/default.nix
index 7ba0ce7d438..c92ece7353b 100644
--- a/pkgs/applications/version-management/forgejo/default.nix
+++ b/pkgs/applications/version-management/forgejo/default.nix
@@ -12,6 +12,12 @@
 , sqliteSupport ? true
 }:

+let
+  tags = lib.optional pamSupport "pam"
+    ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
+  tagsString = lib.concatStringsSep " " tags;
+in
+
 buildGoPackage rec {
   pname = "forgejo";
   version = "1.18.0-1";
@@ -35,18 +41,10 @@ buildGoPackage rec {

   buildInputs = lib.optional pamSupport pam;

-  preBuild =
-    let
-      tags = lib.optional pamSupport "pam"
-        ++ lib.optional sqliteSupport "sqlite sqlite_unlock_notify";
-      tagsString = lib.concatStringsSep " " tags;
-    in
-    ''
-      export buildFlagsArray=(
-        -tags="${tagsString}"
-        -ldflags='-X "main.Version=${version}" -X "main.Tags=${tagsString}"'
-      )
-    '';
+  ldflags = [
+    "-X main.Version=${version}"
+    "-X main.Tags=${tagsString}"
+  ];

   outputs = [ "out" "data" ];

```